std/experimental/color/rgb.d

```diff
@@ -0,0 +1,525 @@
+module std.experimental.color.rgb;
+
+import std.experimental.color;
+import std.experimental.color.conv: convertColor;
+
+import std.traits: isInstanceOf, isNumeric, isIntegral, isFloatingPoint, isSigned, isSomeChar, Unqual;
+import std.typetuple: TypeTuple;
+import std.typecons: tuple;
+
+@safe: pure: nothrow: @nogc:
+
+
+enum isRGB(T) = isInstanceOf!(RGB, T);
+
+
+// DEBATE: which should it be?
+template defaultAlpha(T)
+{
+/+
+    enum defaultAlpha = isFloatingPoint!T ? T(1) : T.max;
++/
+    enum defaultAlpha = T(0);
+}
+
+
+enum RGBColorSpace
+{
+    sRGB,
+    sRGB_Gamma2_2,
+
+/+ TODO: these all need testing
+    AdobeRGB,
+    AppleRGB,
+    BestRGB,
+    BetaRGB,
+    BruceRGB,
+    CIE_RGB,
+    ColorMatchRGB,
+    DonRGB4,
+//    ECI_RGB_v2,
+    EktaSpacePS5,
+    NTSC_RGB,
+    PAL_SECAM_RGB,
+    ProPhotoRGB,
+    SMPTE_C_RGB,
+    WideGamutRGB
++/
+
+    // custom color space will disable automatic color spoace conversions
+    Custom = -1
+}
+
+
+struct RGB(string components_, ComponentType_, bool linear_ = false, RGBColorSpace colorSpace_ = RGBColorSpace.sRGB) if(isValidComponentType!ComponentType_)
+{
+@safe: pure: nothrow: @nogc:
+
+    // RGB colors may only contain components 'rgb', or 'l' (luminance)
+    // They may also optionally contain an 'a' (alpha) component, and 'x' (unused) components
+    static assert(allIn!("rgblax", components), "Invalid Color component '"d ~ notIn!("rgblax", components) ~ "'. RGB colors may only contain components: r, g, b, l, a, x"d);
+    static assert(anyIn!("rgbal", components), "RGB colors must contain at least one component of r, g, b, l, a.");
+    static assert(!canFind!(components, 'l') || !anyIn!("rgb", components), "RGB colors may not contain rgb AND luminance components together.");
+
+    // create members for some useful information
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum RGBColorSpace colorSpace = colorSpace_;
+    enum bool linear = linear_;
+
+
+    // mixin will emit members for components
+    template Components(string components)
+    {
+        static if(components.length == 0)
+            enum Components = "";
+        else
+            enum Components = ComponentType.stringof ~ ' ' ~ components[0] ~ " = 0;\n" ~ Components!(components[1..$]);
+    }
+    mixin(Components!components);
+
+    enum bool hasComponent(char c) = mixin("is(typeof(this."~c~"))");
+    enum bool hasAlpha = hasComponent!'a';
+
+
+    // functions that return the color channels as a tuple
+    @property auto tristimulus() const
```

Comment:
Perhaps I should just define a struct instead of returning a tuple? Although I'll definitely document this detail.